crates/typst/src/visualize/color.rs

```diff
@@ -1130,6 +1130,72 @@ impl Color {
     ) -> StrResult<Color> {
         Self::mix_iter(colors, space)
     }
+
+    /// Make a color more transparent by a given factor.
+    ///
+    /// This method is relative to the existing alpha value.
+    ///
+    /// ```example
+    /// #block(fill: red)[opaque]
+    /// #block(fill: red.transparentize(50%))[half red]
+    /// #block(fill: red.transparentize(75%))[quarter red]
+    /// ```
+    #[func]
+    pub fn transparentize(
+        self,
+        /// The factor to change the alpha value by.
+        factor: Ratio,
+    ) -> StrResult<Color> {
+        #[inline]
+        fn transform<C>(mut color: Alpha<C, f32>, factor: Ratio) -> Alpha<C, f32> {
+            color.alpha = (color.alpha * (1.0 - factor.get() as f32)).clamp(0.0, 1.0);
+            color
+        }
+
+        Ok(match self {
+            Color::Luma(c) => Color::Luma(transform(c, factor)),
+            Color::Oklab(c) => Color::Oklab(transform(c, factor)),
+            Color::Oklch(c) => Color::Oklch(transform(c, factor)),
+            Color::Rgb(c) => Color::Rgb(transform(c, factor)),
+            Color::LinearRgb(c) => Color::LinearRgb(transform(c, factor)),
+            Color::Cmyk(_) => bail!("CMYK does have an alpha component"),
+            Color::Hsl(c) => Color::Hsl(transform(c, factor)),
+            Color::Hsv(c) => Color::Hsv(transform(c, factor)),
+        })
+    }
+
+    /// Make a color more opaque by a given factor.
+    ///
+    /// This method is relative to the existing alpha value.
+    ///
+    /// ```example
+    /// #block(fill: red)[opaque]
+    /// #block(fill: red.opacify(50%))[half red]
+    /// #block(fill: red.opacify(75%))[quarter red]
+    /// ```
+    #[func]
+    pub fn opacify(
+        self,
+        /// The factor to change the alpha value by.
+        factor: Ratio,
+    ) -> StrResult<Color> {
+        #[inline]
+        fn transform<C>(mut color: Alpha<C, f32>, factor: Ratio) -> Alpha<C, f32> {
+            color.alpha = (color.alpha * (1.0 + factor.get() as f32)).clamp(0.0, 1.0);
```

Comment:
This is not in line with what Sass recommends for their `color.scale` method: They take the remainder between the current alpha and a totally opaque color and scale that, so that would work like this:
```suggestion
            color.alpha = (factor.get() as f32 * (1.0 - color.alpha) + color.alpha).clamp(0.0, 1.0);
```

Reply:
I think I agree with this choice. It gets away with the meaning of "opacify(50%) means that the color is now 50% more opaque compared to before", but that is for a good cause.